hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1387,45 +1387,82 @@ def should_replace_byte(b):
+                        return False
+                    else:
+                        attempt[u:v] = int_to_bytes(value - 1, v - u)
+                elif d == "X":
+                    del attempt[u:v]
+                else:  # pragma: no cover
+                    assert False, "Unrecognised command %r" % (d,)
+        return self.incorporate_new_buffer(attempt)
+
 
 def block_program(description):
-    """Mini-DSL for block rewriting. A sequence of commands that will be run
-    over all contiguous sequences of blocks of the description length in order.
-    Commands are:
-
-        * ".", keep this block unchanged
-        * "-", subtract one from this block.
-        * "0", replace this block with zero
-        * "X", delete this block
-
-    If a command does not apply (currently only because it's - on a zero
-    block) the block will be silently skipped over. As a side effect of
-    running a block program its score will be updated.
-    """
+    """Defines a shrink pass that runs the block program ``description``
+    at every block index."""
+    n = len(description)
 
     def generate_arguments(self):
-        return [(i,) for i in hrange(len(self.blocks))]
+        return [(i,) for i in hrange(len(self.blocks) - n)]
 
     def run(self, i):
-        n = len(description)
-        if i + n > len(self.shrink_target.blocks):
+        """Adaptively attempt to run the block program at the current
+        index. If this successfully applies the block program ``k`` times
+        then this runs in ``O(log(k))`` test function calls."""
+        assert i + n < len(self.shrink_target.blocks)
+        # Because we run in a random order we will often find ourselves in the middle
+        # of a region where we could run the block program. We thus start by moving
+        # left to the beginning of that region if possible in order to to start from
+        # the beginning of that region..
+        move_left = find_integer(
+            lambda k: self.run_block_program(
+                i + (1 - k) * n, description, original=self.shrink_target
+            )
+        )
+
+        # If move_left is 0 then we cannot run the block program once at the current
+        # index, so we stop here.
+        if move_left == 0:
             return
-        attempt = bytearray(self.shrink_target.buffer)
-        for k, d in reversed(list(enumerate(description))):
-            j = i + k
-            block = self.blocks[j]
-            u, v = block.bounds
-            if d == "-":
-                value = int_from_bytes(attempt[u:v])
-                if value == 0:
-                    return
-                else:
-                    attempt[u:v] = int_to_bytes(value - 1, block.length)
-            elif d == "X":
-                del attempt[u:v]
-            else:  # pragma: no cover
-                assert False, "Unrecognised command %r" % (d,)
-        self.incorporate_new_buffer(attempt)
+
+        i -= move_left * n - 1
```

Comment:
This doesn't make sense to me. Is it supposed to be `i -= (move_left - 1) * n`?

Reply:
It is. I've added a common helper function to make it impossible to mess that up.